### PR TITLE
Changes for CircleCI

### PR DESCRIPTION
**Pull request by isaac-s**

### CODE REVIEW

File: resources/cfyci.py [hunk 1]
```diff
@@ -0,0 +1,602 @@
...
+# When we initialize the profile (in the initialize() function), we must do so
+# with the tenant name injected into the profile; we must not leave it empty.
+# Otherwise, REST calls would fail because @pass_client wouldn't pass a tenant
+# to the REST call.
+#
+# However, if we inject the tenant to the profile, and the "CLOUDIFY_TENANT" environment
+# variable is defined, then we get an error message that the tenant is defined
+# in both the profile and an environment variable.
+#
+# To work around that, I invented a new environment variable.
```
Comment (alexmolev) [inline resources/cfyci.py:60]: It's not a place to write it,
If there is a BUG/Feature Request please open Jira ticket.

The documentation part should be in the documentation.

File: resources/cfyci.py [hunk 2]
```diff
@@ -0,0 +1,602 @@
...
+
+
+def read_json_or_yaml(path):
+    # We assume here, of course, that any JSON file is also a YAML file.
+    with open(path, 'r') as f:
+        return yaml.load(f)
+
+
+def initialize(**kwargs):
+    # Calls the CLI to set a profile. This should normally only happen once.
```
Comment (alexmolev) [inline resources/cfyci.py]: please remove the comment

File: resources/cfyci.py [hunk 3]
```diff
@@ -0,0 +1,602 @@
...
+    if inputs:
+        if type(inputs) == dict:
+            with tempfile.NamedTemporaryFile(suffix=".yaml", delete=False) as temp_inputs_file:
+                logger.info("Created temporary file for inputs: %s", temp_inputs_file.name)
+                yaml.safe_dump(inputs, temp_inputs_file)
+                inputs_file_name = temp_inputs_file.name
+        elif type(inputs) == str:
+            inputs_file_name = inputs
+        else:
+            raise Exception("Unhandled inputs type: %s" % type(inputs))
```
Comment (alexmolev) [inline resources/cfyci.py:154]: specify what are the expected type.

File: resources/cfyci.py [hunk 4]
```diff
@@ -0,0 +1,602 @@
...
+        cmdline.extend(['-i', inputs_file_name])
+    try:
+        subprocess.check_call(cmdline)
+    finally:
+        if temp_inputs_file:
+            logger.info("Deleting temporary file: %s", temp_inputs_file.name)
+            os.remove(temp_inputs_file.name)
+    # Now wait until the deployment deletion ended.
+    # Since there's no way to get the execution ID of the deployment creation,
+    # we need to look it up (see https://cloudifysource.atlassian.net/browse/CY-2385).
```
Comment (alexmolev) [inline resources/cfyci.py:164]: please remove the comments

File: resources/cfyci.py [hunk 5]
```diff
@@ -0,0 +1,602 @@
...
+    executions = client.executions.list(deployment_id=name)
+    if len(executions) != 1:
+        raise Exception("Unexpected number of executions for deployment '%s': %d" % (name, len(executions)))
```
Comment (alexmolev) [inline resources/cfyci.py:167]: specify that the expected number is 1

File: resources/cfyci.py [hunk 6]
```diff
@@ -0,0 +1,602 @@
...
+    execution = wait_for_execution(
+        client, executions[0], get_events_logger(False), True, timeout=None, logger=logger)
+    if execution.status != Execution.TERMINATED:
+        raise Exception("Unexpected status of execution %s: %s" % (execution.id, execution.status))
```
Comment (alexmolev) [inline resources/cfyci.py:171]: Specify what is expected status

File: resources/cfyci.py [hunk 7]
```diff
@@ -0,0 +1,602 @@
...
+
+
+def _start_and_follow_execution(client, deployment_id, workflow_id, parameters):
+    # Use REST here, because "cfy executions start" ends with a zero
+    # even if the execution fails.
```
Comment (alexmolev) [inline resources/cfyci.py:176]: remove the comments

File: resources/cfyci.py [hunk 8]
```diff
@@ -0,0 +1,602 @@
...
+    execution = client.executions.start(deployment_id, workflow_id, parameters)
+    execution = wait_for_execution(client, execution, get_events_logger(False), True, timeout=None, logger=logger)
+    if execution.status != Execution.TERMINATED:
+        raise Exception("Unexpected status of execution %s: %s" % (execution.id, execution.status))
```
Comment (alexmolev) [inline resources/cfyci.py:180]: Looks like duplicated code of the lines 121-124 and it looks like a pattern, let's move it to a method

File: resources/cfyci.py [hunk 9]
```diff
@@ -0,0 +1,602 @@
...
+    })
+
+
+@pass_client()
+def _delete_deployment(name, client):
+    subprocess.check_call([
+        'cfy', 'deployments', 'delete', name
+    ])
+    # Wait until the deployment is actually deleted, as "cfy deployments delete"
+    # is asynchronous.
```
Comment (alexmolev) [inline resources/cfyci.py:201]: remove comments

File: resources/cfyci.py [hunk 10]
```diff
@@ -0,0 +1,602 @@
...
+    while True:
```
Comment (alexmolev) [inline resources/cfyci.py:202]: What happens if the deletion has failed and the deployment is stuck?
Let's add some TIMEOUT mechanism with a default value that can be adjusted through environment variable

File: resources/cfyci.py [hunk 11]
```diff
@@ -0,0 +1,602 @@
...
+        # kwargs.pop('func', None)
+        # invocation_params_file = kwargs.pop("invocation_params_file", None)
+        # invocation_params = {}
+        # if invocation_params_file:
+        #     invocation_params = read_json_or_yaml(invocation_params_file)
+        # for arg_name, arg_value in kwargs.iteritems():
+        #     if arg_value is not None:
+        #         invocation_params[arg_name] = arg_value
+        # print(invocation_params)
+        # func(*args, **invocation_params)
```
Comment (alexmolev) [inline resources/cfyci.py:250]: please clean it up

File: resources/cfyci.py [hunk 12]
```diff
@@ -0,0 +1,602 @@
...
+        func(*args, **kwargs)
+
+    return wrapper
+
+
+def create_deployment(name, blueprint, inputs_file, **kwargs):
+    _create_deployment(name, blueprint, inputs_file)
```
Comment (alexmolev) [inline resources/cfyci.py:257]: why just not make create deployment public?

File: resources/cfyci.py [hunk 13]
```diff
@@ -0,0 +1,602 @@
...
+    def __init__(self, name, outputs_file, module, variables, environment, environment_mapping):
+        CfyIntegration.__init__(self, name, outputs_file)
+        self._module = module
+        self._variables = variables
+        self._environment = environment
+        self._environment_mapping = CfyIntegration.parse_environment_mapping(
+            environment_mapping)
+
+    def blueprint_name(self):
+        return 'cfy-terraform-1.0'
```
Comment (alexmolev) [inline resources/cfyci.py:321]: why is the blueprint is a fixed name?
If it should be a fixed name let's make a global const environment and put it at the top.

**Review by alexmolev — APPROVED** [review]

File: resources/cfyci.py (hunk 3 above)
Comment (isaac-s) [reply resources/cfyci.py:154]: Will be done in a follow-up PR

File: resources/cfyci.py (hunk 1 above)
Comment (isaac-s) [reply resources/cfyci.py:60]: > The documentation part should be in the documentation.

I'll move it to `README.md`

File: resources/cfyci.py (hunk 4 above)
Comment (isaac-s) [reply resources/cfyci.py:164]: So where do you expect this comment to be? It shouldn't be removed. It's an important marker for future improvement.

File: resources/cfyci.py (hunk 8 above)
Comment (isaac-s) [reply resources/cfyci.py:180]: Will be done in a follow-up PR

File: resources/cfyci.py (hunk 10 above)
Comment (isaac-s) [reply resources/cfyci.py:202]: Follow-up PR

File: resources/cfyci.py (hunk 13 above)
Comment (isaac-s) [reply resources/cfyci.py:321]: It is a fixed name in order to align with the same expectations we have with the Jenkins plugin.
There's an ongoing pending item how to approach this more globally.
I'll change this to constants.

**Pull request merged**